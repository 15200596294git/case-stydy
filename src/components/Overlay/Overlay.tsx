import {
  defineComponent,
  PropType,
  CSSProperties,
  ExtractPropTypes,
  WatchSource,
  Transition,
} from 'vue'

import { useEventListener } from '@vueuse/core'

import { isDef, extend, preventDefault } from '@/utils'
import { getZIndexStyle } from '@/utils/format'
import { createNamespace } from '@/utils/create'

import '../index.less'

const numericProp = [Number, String]

const truthProp = {
  type: Boolean,
  // 强制将类型变为不可变类型
  default: true as const,
}

const unknownProp = null as unknown as PropType<unknown>

const [name, bem] = createNamespace('overlay')

export const overlayProps = {
  show: Boolean,
  zIndex: numericProp,
  duration: numericProp,
  className: unknownProp,
  lockScroll: truthProp,
  lazyRender: truthProp,
  customStyle: Object as PropType<CSSProperties>,
}

export type OverlayProps = ExtractPropTypes<typeof overlayProps>

// 传入一个值，该值的类型为watch函数监听的类型
// 返回一个渲染函数
// 如果传入的值为true，立马执行渲染函数
// 否则会监听传入的值，直到该值变为true后，执行渲染函数
const useLazyRender = (show: WatchSource<boolean | undefined>) => {
  const inited = ref(false)

  watch(
    show,
    (value) => {
      if (value) {
        inited.value = true
      }
    },
    { immediate: true },
  )

  return (render: () => JSX.Element) => (inited.value ? render() : null)
}

export default defineComponent({
  name: 'cs-overlay',

  props: overlayProps,

  setup(props, { slots }) {
    const lazyRender = useLazyRender(() => props.show || !props.lazyRender)

    const onTouchMove = (event: TouchEvent) => {
      if (props.lockScroll) {

        preventDefault(event, true)
        console.log('aaa');
        return false
      }
    }

    return () => {
      const root = ref<HTMLDivElement>()
      useEventListener(root, 'touchmove', onTouchMove, {
        capture: false,
        passive: false,
      })

      const renderOverlay = lazyRender(() => {
        const style: CSSProperties = extend(
          getZIndexStyle(props.zIndex),
          props.customStyle,
        )
        if (isDef(props.duration)) {
          style.animationDuration = `${props.duration}s`
        }

        return (
          <div
            v-show={props.show}
            ref={root}
            style={style}
            class={[bem(), props.className]}
          >
            {slots.default?.()}
          </div>
        )
      })

      return (
        <Transition
          v-slots={{ default: renderOverlay }}
          name="cs-fade"
          appear
        ></Transition>
      )
    }
  },
})
