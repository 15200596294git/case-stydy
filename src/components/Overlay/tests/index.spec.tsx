import { test, expect, vi } from 'vitest'

import { mount } from '@vue/test-utils'

import Overlay from '../Overlay'

test('传入z-index属性，组件的z-index应该和传入的z-index属性相等', () => {
  const wrapper = mount(Overlay, {
    props: {
      show: true,
      zIndex: 99,
    },
  })

  const overlay = wrapper.find('.cs-overlay')

  expect((overlay.element as HTMLElement).style.zIndex).toBe('99')
})

test('可以传入class-name prop，来自定义class', () => {
  const wrapper = mount(Overlay, {
    props: {
      show: true,
      className: 'foo',
    },
  })

  const overlay = wrapper.find('.cs-overlay')
  expect(overlay.classes()).toContain('foo')
})

test('可以传入custom-style prop，来自定义style', () => {
  const wrapper = mount(Overlay, {
    props: {
      show: true,
      customStyle: {
        backgroundColor: 'red',
      },
    },
  })

  const overlay = wrapper.find('.cs-overlay')
  expect((overlay.element as HTMLElement).style.backgroundColor).toEqual('red')
})

test('传入duration prop，来改变默认的动画duration', () => {
  const wrapper = mount(Overlay, {
    props: {
      show: true,
      duration: 1,
    },
  })

  const overlay = wrapper.find('.cs-overlay')
  expect((overlay.element as HTMLElement).style.animationDuration).toEqual('1s')
})

test('应该正确的渲染传入的插槽', () => {
  const wrapper = mount(Overlay, {
    slots: {
      default: () => 'Custom Default',
    },
  })

  // 第一次运行会生成一个快照
  // 第二次运行时会和已经生成的快照对比，如果不匹配测试就不会通过
  // 否则测试通过
  expect(wrapper.html()).toMatchSnapshot()
})

test('传入的lock-scroll prop为false时，应该能触发touchMove事件', () => {
  const onTouchMove = vi.fn()
  const wrapper = mount({
    render() {
      return (
        <div onTouchmove={onTouchMove}>
          <Overlay show lockScroll={false}/>
        </div>
      )
    },
  })

  const overlay = wrapper.find('.cs-overlay')
  overlay.trigger('touchmove')
  expect(onTouchMove).toHaveBeenCalledTimes(1)
})

// test('传入的lock-scroll prop为true时，应该不能触发touchMove事件', () => {
//   const onTouchMove = vi.fn()
//   const wrapper = mount({
//     render() {
//       return (
//         <div onTouchmove={onTouchMove}>
//           <Overlay show lockScroll />
//         </div>
//       )
//     },
//   })

//   const overlay = wrapper.find('.cs-overlay')
//   overlay.trigger('touchmove')
//   expect(onTouchMove).toHaveBeenCalledTimes(0)
// })

test('lazyRender为false时，禁用懒加载', () => {
  const wrapper = mount({
    render() {
      return <Overlay lazyRender={false} />
    },
  })

  expect(wrapper.html()).toMatchSnapshot()
})
