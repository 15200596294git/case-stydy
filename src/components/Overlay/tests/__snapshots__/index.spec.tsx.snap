// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lazyRender为false时，禁用懒加载 1`] = `
"<transition-stub name=\\"cs-fade\\" appear=\\"true\\" persisted=\\"false\\" css=\\"true\\">
  <div class=\\"cs-overlay\\" style=\\"display: none;\\">
    <!---->
  </div>
</transition-stub>"
`;

exports[`应该正确的渲染传入的插槽 1`] = `"<transition-stub name=\\"cs-fade\\" appear=\\"true\\" persisted=\\"false\\" css=\\"true\\"></transition-stub>"`;
